### PR TITLE
Add GPT2 and Update Github Actions

### PR DESCRIPTION
**Pull request by rsuderman**
These changes make a few large changes. It adds GPT-2 with scripts for performing end-to-end tests
with numerical validation. It also includes updates to the github actions so that we run verification that:

- Unittests are passing (on pull)
- Pinned versions are automagically updated (nightly)
- GPT-2 is still working (nightly)

GPT-2 sample run executed below:
https://github.com/rsuderman/iree-jax/runs/6264872065?check_suite_focus=true

### CODE REVIEW

File: models/gpt2/model.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
+from os import path
+import functools
```
Comment (ScottTodd) [inline models/gpt2/model.py:2]: Please add license headers to all source files

Also check that the license header style is consistent across the project. I see two different versions:

https://github.com/google/iree-jax/blob/b26ae072cc80f8d2fa9e90c26d790d3b3a624c03/setup.py#L1-L5

https://github.com/google/iree-jax/blob/b26ae072cc80f8d2fa9e90c26d790d3b3a624c03/iree/jax/builtins.py#L1-L13

(and it looks like iree-jax doesn't have the LLVM exception set right now: https://github.com/google/iree-jax/blob/main/LICENSE, compared to https://github.com/google/iree/blob/main/LICENSE, so the longer form might be correct?)

File: .github/workflows/release_wheel.yaml [hunk 2]
```diff
@@ -15,7 +15,7 @@ jobs:
       - uses: actions/setup-python@v2
```
Comment (ScottTodd) [inline .github/workflows/release_wheel.yaml:15]: Technically google org policy is now to pin to a specific version (`uses: actions/setup-python@7f80679172b057fc5e90d70d197929d454754a5a  # v2`), not a tagged release (`uses: actions/setup-python@v2`): https://opensource.google/documentation/reference/github/services#actions

I just updated the main IREE repo: https://github.com/google/iree/pull/8997 the other day. The risk area is smaller in this repo, but would still be nice to fix at some point.

File: .github/workflows/test_gpt2_model.yaml [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        run: |
+          models/gpt2/setup.sh
+
+      - name: Export GPT-2
+        run: |
+          python $GITHUB_WORKSPACE/models/gpt2/export.py
+
+      - name: Test GPT-2
+        run: |
+          python $GITHUB_WORKSPACE/models/gpt2/test_export.py
```
Comment (ScottTodd) [inline .github/workflows/test_gpt2_model.yaml:40]: How long does this take on average? Just a few seconds/minutes? If larger, we'd want to be careful about using up too much time on the actions runners, since they are a shared resource throughout the org.

Looks like https://github.com/rsuderman/iree-jax/runs/6264872065?check_suite_focus=true is a sample run? (would be nice to include that in the PR description)

File: .github/workflows/test_head_and_bump.yaml [hunk 4]
```diff
@@ -12,15 +12,33 @@ jobs:
...
+        run: |
+          python build_tools/update_pinned.py
+          cat version_info.json
+
+      - name: Push updated versions
+        run: |
+          git config --global user.name 'Rob Suderman'
+          git config --global user.email 'rsuderman@users.noreply.github.com'
+          git commit -am "Update pinned versions"
+          git push
```
Comment (ScottTodd) [inline .github/workflows/test_head_and_bump.yaml:44]: nit: files should end in a newline character

File: .github/workflows/test_head_and_bump.yaml [hunk 5]
```diff
@@ -12,15 +12,33 @@ jobs:
...
-      # version_info.json and push.
+      - name: Update Pinned Versions
+        run: |
+          python build_tools/update_pinned.py
+          cat version_info.json
+
+      - name: Push updated versions
+        run: |
+          git config --global user.name 'Rob Suderman'
+          git config --global user.email 'rsuderman@users.noreply.github.com'
```
Comment (ScottTodd) [inline .github/workflows/test_head_and_bump.yaml:42]: Probably want to use a bot name + email for this (in the common repo, you can do whatever in your fork)

File: build_tools/configure.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+  version_info_file = os.path.join(project_root, "version_info.json")
+  try:
+    with open(version_info_file, "rt") as f:
+      version_info = json.load(f)
+  except FileNotFoundError:
+    print >> sys.stderr, "version_info.json not found"
+    sys.exit(1)
+
+  args = parser.parse_args()
+  version_info["use-pinned"] = True if args.use_pinned else False
```
Comment (ScottTodd) [inline build_tools/configure.py:39]: Hmm, this is a little weird. I guess `BooleanOptionalAction` can be `None`, `True`, or `False`? Is that more intuitive than a boolean with a default value?

File: build_tools/configure.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    del version_info["package-version"]
+
+  print(json.dumps(version_info, indent=2))
+
+  with open(os.path.join(project_root, "version_info.json"), 'w') as f:
+      f.write(json.dumps(version_info, indent=2))
+      f.write("\n")
+
+if __name__ == "__main__":
+  main()
```
Comment (ScottTodd) [inline build_tools/configure.py:53]: Can you add a docstring to the top of this file explaining what it configures, when it should be ran, etc.? Is it just for automation, or are users expected to run it at some point?

File: build_tools/update_pinned.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        pinned[a] = reqs[a]
+
+  version_info = {a : version_info[a] for a in version_info if a == "pinned-versions"}
+
+  with open(os.path.join(project_root, "version_info.json"), 'w') as f:
+      f.write(json.dumps(version_info, indent=2))
+      f.write("\n")
+
+if  __name__ == "__main__":
+  main()
```
Comment (ScottTodd) [inline build_tools/update_pinned.py:48]: docstring for the file please

(and nit: files should end with a newline character)

File: build_tools/update_pinned.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    with open(version_info_file, "rt") as f:
+      version_info = json.load(f)
+  except FileNotFoundError:
+    print >> sys.stderr, "version_info.json not found"
+    sys.exit(1)
+
+  pinned = version_info["pinned-versions"]
+  for a in pinned:
+      if a in reqs:
+        pinned[a] = reqs[a]
```
Comment (ScottTodd) [inline build_tools/update_pinned.py:39]: nit: more descriptive variable name?

File: models/gpt2/config.py [hunk 10]
```diff
@@ -0,0 +1,17 @@
+'''Flags and model configuration details, useful for replicating across files.'''
+from absl import flags
+from os import path
+
+import collections
+import pathlib
+
+assets_dir = path.join(pathlib.Path(__file__).parent.resolve(), "assets")
+
+flags.DEFINE_string('binary_path', '/tmp/gpt2.vfmb', 'Path for binary')
```
Comment (ScottTodd) [inline models/gpt2/config.py:10]: typo: vmfb

File: models/gpt2/export.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+import config
+import model
+
+from iree.jax import Program, store_global
+from jax.core import ShapedArray
+from os import path
+
+FLAGS = absl.flags.FLAGS
+
+def CreateGpt2Model(name, B, K, S, T):
```
Comment (ScottTodd) [inline models/gpt2/export.py:21]: Can you document what `B`, `K`, `S`, and `T` are? Maybe like to a page or paper talking about the model architecture.

File: models/gpt2/model.py [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+import jax
+import jax.numpy as jnp
+from jax._src import abstract_arrays
+import numpy as np
+
+
+# pylint: disable=invalid-name
+def load_gpt2_model(name, gpt2_dir):
+  """Load GPT-2 pretrained weights into JAX DeviceArrays."""
+  # TODO(jekbradbury): add target dtype
```
Comment (ScottTodd) [inline models/gpt2/model.py:14]: 🤔 change the TODO name or update attribution for this file?

File: .github/workflows/test_presubmit.yaml [hunk 13]
```diff
@@ -15,12 +15,20 @@ jobs:
...
+          python3 -m pip install Filecheck
+
+      - name: Run installation
+        run: |
+          python $GITHUB_WORKSPACE/build_tools/configure.py --use-pinned
+          python3 -m pip install . -f https://github.com/google/iree/releases --force-reinstall .[xla,cpu,test]
+      
+      - name: Run tests
+        run: |
+          build_tools/unittest.sh
```
Comment (ScottTodd) [inline .github/workflows/test_presubmit.yaml:31]: These "update github actions" changes and the unittest script seem unrelated to the GPT2 changes? Could that be split into at least two commits, if not two fully separate PRs? Specifically I'm looking at `build_tools/unittest.sh` and wondering why `models/gpt2/test_jax.py` (or other python tests) is not included there (why have a script that just calls lit?)

**Review by ScottTodd — CHANGES_REQUESTED** [review]

File: .github/workflows/test_gpt2_model.yaml (hunk 3 above)
Comment (rsuderman) [reply .github/workflows/test_gpt2_model.yaml:40]: About two minutes, it should be short enough that a nightly should be fine. It may be possible to run it only in cases where a version update occurs which is the main goal.

File: .github/workflows/test_presubmit.yaml (hunk 13 above)
Comment (rsuderman) [reply .github/workflows/test_presubmit.yaml:31]: I needed to change how configuration worked selecting versions and adding GPT-2. Potentially I could remove all the GPT-2 work and commit that in a separate follow up but the original inspiration to these changes was adding GPT-2.

File: build_tools/configure.py (hunk 7 above)
Comment (rsuderman) [reply build_tools/configure.py:59]: Done.

File: models/gpt2/export.py (hunk 11 above)
Comment (rsuderman) [reply models/gpt2/export.py:26]: Done along with replicated in config.py

File: models/gpt2/model.py (hunk 1 above)
Comment (rsuderman) [reply models/gpt2/model.py:2]: Changed to consistent version and used long form version.

File: build_tools/update_pinned.py (hunk 9 above)
Comment (rsuderman) [reply build_tools/update_pinned.py]: Done.

File: models/gpt2/config.py (hunk 10 above)
Comment (rsuderman) [reply models/gpt2/config.py]: Done.

File: models/gpt2/model.py (hunk 12 above)
Comment (rsuderman) [reply models/gpt2/model.py]: Removed, was carry over from the originally demo code.

File: .github/workflows/test_head_and_bump.yaml (hunk 4 above)
Comment (rsuderman) [reply .github/workflows/test_head_and_bump.yaml]: Done.

File: .github/workflows/test_head_and_bump.yaml (hunk 5 above)
Comment (rsuderman) [reply .github/workflows/test_head_and_bump.yaml]: Done.

File: .github/workflows/release_wheel.yaml (hunk 2 above)
Comment (rsuderman) [reply .github/workflows/release_wheel.yaml]: Done.

**Review by ScottTodd — APPROVED** [review]
